internal/app/upstream/client.go

```diff
@@ -55,11 +74,179 @@ type Response struct {
+// - SendMsg is performed with timeout.
+// - send is a receiver for signal and exits when signal is closed by the owner.
+// - send also exits on context cancellations.
+func send(
+	ctx context.Context,
+	logger log.Logger,
+	cancelFunc context.CancelFunc,
+	request *v2.DiscoveryRequest,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(ctx, func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					// Context was cancelled, hence this is not an erroneous scenario.
+				default:
+					logger.Error(ctx, "Error in SendMsg: %s", err.Error())
+				}
+				cancelFunc()
+				return
+			}
+		case <-ctx.Done():
+			_ = stream.CloseSend()
+			return
+		}
+	}
+}
+
+// recv is an infinite loop which blocks on RecvMsg.
+// The only ways to exit the goroutine is by cancelling the context or when an error occurs.
+func recv(
+	ctx context.Context,
+	cancelFunc context.CancelFunc,
+	logger log.Logger,
+	response chan *v2.DiscoveryResponse,
+	stream grpc.ClientStream,
+	signal chan *version) {
+	for {
+		resp := new(v2.DiscoveryResponse)
+		if err := stream.RecvMsg(resp); err != nil {
+			select {
+			case <-ctx.Done():
+				// Context was cancelled, hence this is not an erroneous scenario.
+			default:
+				logger.Error(ctx, "Error in RecvMsg %s", err.Error())
+			}
+			cancelFunc()
+			closeChannels(signal, response)
+			return
+		}
+		select {
+		case <-ctx.Done():
+			closeChannels(signal, response)
+			return
+		default:
+			response <- resp
+			signal <- &version{version: resp.GetVersionInfo(), nonce: resp.GetNonce()}
+		}
+	}
+}
+
+// closeChannels is called whenever the context is cancelled (ctx.Done) in Send and Recv goroutines.
+// It is also called when an irrecoverable error occurs and the error is passed to the caller.
+func closeChannels(versionChan chan *version, responseChan chan *v2.DiscoveryResponse) {
+	close(versionChan)
+	close(responseChan)
+}
+
+// shutDown should be called in a separate goroutine.
+// This is a blocking function that closes the upstream connection on context completion.
+func shutDown(ctx context.Context, conn *grpc.ClientConn) {
+	<-ctx.Done()
+	conn.Close()
+}
+
+// DoWithTimeout runs f and returns its error.  If the deadline d elapses first,
+// it returns a DeadlineExceeded error instead.
+// Ref: https://github.com/grpc/grpc-go/issues/1229#issuecomment-302755717
+func doWithTimeout(ctx context.Context, f func() error, d time.Duration) error {
+	deadlineCtx, cancel := context.WithDeadline(ctx, time.Now().Add(d))
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- f()
+		close(errChan)
+	}()
+	select {
+	case <-deadlineCtx.Done():
+		cancel()
```

Comment:
This cancel context is not passed anywhere..not sure what this does.

Reply:
i was flummoxed by this error message. Since the cancel operation is idempotent, and `deadlineCtx.Done()` basically means the deadline context is already cancelled, i used the cancel to be a noop and also to get rid of the error.
<img width="587" alt="Screen Shot 2020-04-17 at 9 00 31 AM" src="https://user-images.githubusercontent.com/49211422/79589759-c9efce80-808a-11ea-9e7c-f98be689d146.png">

Reply:
I see. If you are calling cancel, I would just `defer cancel()` so you don't have to call it everywhere. `cancel()` is needed to release the resources it is associated with so it cleans itself up.

Also, does it make sense to use context.WithTimeout so it sends a timeout error? If not, then I would rename the private method to be `doWithDeadline`.

Reply:
`context.WithTimeout` makes more sense. Will change `WithDeadline` to `WithTimeout`